### PR TITLE
space-age: improve exercise implementation

### PR DESCRIPTION
**Pull request by dscottboggs**
I believe this implementation more accurately reflects the statically-typed nature of the language, and hopefully teaches the learner some of Crystal's macro primitives.

### CODE REVIEW

**Comment by mhelmetag** [comment]
Hey there, thanks for the PR!

We're trying not to edit the tests directly. Basically they're generated based with canonical test data (https://github.com/exercism/problem-specifications) and a generator (https://github.com/exercism/crystal/blob/master/generator/src/generators/space_age.cr).

Also, using macros is a great way to solve this (vs a hash)... however, these "examples" are basically just ways to test the test and generator. If people want examples, they should be looking at the ones in exercism.

Does that make sense?

I'd be more than happy to merge a change that includes a change to the test generator but, as is, I believe this goes against what we've been doing (having generators to automagically update the specs).

**Comment by dscottboggs** [comment]
Thanks for the advice, the generator system looks very nice, much easier to make revisions that way than manually. I'll revise my work. :slightly_smiling_face:

**Comment by dscottboggs** [comment]
I've refactored my changes to use the generator system. Let me know what you think.

File: generator/src/generators/space_age.cr [hunk 1]
```diff
@@ -2,14 +2,15 @@ require "../exercise_generator"
 require "../exercise_test_case"
 
 class SpaceAgeGenerator < ExerciseGenerator
+  record TestCases, cases : Array(SpaceAgeTestCase) do
```
Comment (mhelmetag) [inline generator/src/generators/space_age.cr]: This is pretty cool! Definitely cleaner.

File: exercises/space-age/src/example.cr [hunk 2]
```diff
@@ -1,18 +1,36 @@
-module SpaceAge
-  extend self
+module Space
+  struct Age
+    SECONDS_PER_EARTH_YEAR = 31557600
+    def age_on_mercury
```
Comment (mhelmetag) [inline exercises/space-age/src/example.cr]: Weren't these using a macro to build out the methods before? I thought that sparked the example change?

File: generator/src/generators/space_age.cr [hunk 3]
```diff
@@ -29,7 +30,7 @@ class SpaceAgeTestCase < ExerciseTestCase
   )
 
   def workload
-    "SpaceAge.#{property}(\"#{input.planet}\", #{input.seconds}).should eq(#{expected})"
+    "Space::Age.from_seconds(#{input.seconds}).#{property}_on_#{input.planet.downcase}.should be_close(#{expected}, 0.01)"
```
Comment (mhelmetag) [inline generator/src/generators/space_age.cr]: I do like this more OOP approach. I think that's a good example to have.

File: exercises/space-age/src/example.cr (hunk 2 above)
Comment (dscottboggs) [reply exercises/space-age/src/example.cr]: Yes, but when you said

> Also, using macros is a great way to solve this (vs a hash)... however, these "examples" are basically just ways to test the test and generator. If people want examples, they should be looking at the ones in exercism.

I thought you meant to remove the macros from the example. I'd be glad to put them back.

File: generator/src/generators/space_age.cr (hunk 3 above)
Comment (dscottboggs) [reply generator/src/generators/space_age.cr]: Thanks :slightly_smiling_face:

**Comment by mhelmetag** [comment]
Thanks! Great example of macros, more object oriented-ish programming, and cleaning up our spec generators' JSON handling.

**Pull request merged**